File: javascript/avoid_hadouken_ifs.js
// Avoid Hadouken IFs
// Avoid nested IFs (Hadouken IFs)
// Solution: Early Returns, Switch-Cases

// Before
class Order {
  constructor() {
    this.isValid = true;
    this.isInStock = true;
    this.paymentSuccess = true;
  }

  valid() {
    return this.isValid;
  }

  inStock() {
    return this.isInStock;
  }

  paymentSuccessful() {
    return this.paymentSuccess;
  }
}

function processOrder(order) {
  if (order.valid()) {
    if (order.inStock()) {
      if (order.paymentSuccessful()) {
        console.log("Order processed successfully!");
      } else {
        console.log("Payment failed.");
      }
    } else {
      console.log("Item is out of stock.");
    }
  } else {
    console.log("Order is invalid.");
  }
}

const order = new Order();
processOrder(order);


// After
class Order {
  constructor() {
    this.isValid = true;
    this.isInStock = true;
    this.paymentSuccess = true;
  }

  valid() {
    return this.isValid;
  }

  inStock() {
    return this.isInStock;
  }

  paymentSuccessful() {
    return this.paymentSuccess;
  }
}

function processOrder(order) {
  if (!order.valid()) {
    console.log("Order is invalid.");
    return;
  }
  
  if (!order.inStock()) {
    console.log("Item is out of stock.");
    return;
  }
  
  if (!order.paymentSuccessful()) {
    console.log("Payment failed.");
    return;
  }

  console.log("Order processed successfully!");
}

const order = new Order();
processOrder(order);

File: go/avoid_hadouken_ifs.go
// Avoid Hadouken IFs
// Avoid nested IFs (Hadouken IFs)
// Solution: Early Returns, Switch-Cases

package main

import "fmt"

type Order struct {
	isValid        bool
	isInStock      bool
	paymentSuccess bool
}

func NewOrder() *Order {
	return &Order{
		isValid:        true,
		isInStock:      true,
		paymentSuccess: true,
	}
}

func (o *Order) Valid() bool {
	return o.isValid
}

func (o *Order) InStock() bool {
	return o.isInStock
}

func (o *Order) PaymentSuccessful() bool {
	return o.paymentSuccess
}

// Before
func processOrderBefore(order *Order) {
	if order.Valid() {
		if order.InStock() {
			if order.PaymentSuccessful() {
				fmt.Println("Order processed successfully!")
			} else {
				fmt.Println("Payment failed.")
			}
		} else {
			fmt.Println("Item is out of stock.")
		}
	} else {
		fmt.Println("Order is invalid.")
	}
}

// After
func processOrder(order *Order) {
	if !order.Valid() {
		fmt.Println("Order is invalid.")
		return
	}
	
	if !order.InStock() {
		fmt.Println("Item is out of stock.")
		return
	}
	
	if !order.PaymentSuccessful() {
		fmt.Println("Payment failed.")
		return
	}

	fmt.Println("Order processed successfully!")
}

func main() {
	order := NewOrder()
	
	processOrderBefore(order)
	fmt.Println()
	processOrder(order)
}
